Use Roslyn to create ref/ assemblies

dougbu opened the pull request:
- #14801 

Suggest reviewing the commits individually and mostly ignoring the first ("Remove all ref/ projects"). 

This set of changes
- reduces the complexity of our build, mostly because it removes all ref/ projects
- avoids any need for manual changes when APIs change
- increases the targeting pack size slightly e.g. `Microsoft.AspNetCore.App.Ref.*.nupkg` was about 1.5 MB before this change and 2 MB after
   - most of this comes from including more information in the assemblies e.g. `internal` symbols when IVT is used
- makes it difficult to display public API changes -- until we decide on the baselines to compare against and implement that comparison

dougbu committed:
Remove GenAPI infrastructure

dougbu committed:
Remove all ref/ projects

dougbu committed:
Expose `%(LatestPackageReference.RTMVersion)` metadata
- automate use of properties in the `@(LatestPackageReference)` item group to make this maintainable
  - add a couple of special cases at the bottom of eng/Dependencies.props
  - add one more `$(...PackageVersion)` property to avoid yet-another special case

dougbu committed:
Remove notion of a reference assembly project
- remove `$(IsReferenceAssemblyProject)`, `$(ReferenceReferenceAssemblies)` and `$(ReferenceImplementationAssemblies)`
    - remove unnecessary `$(NoWarn)` settings

nits:
- remove a few misleading comments
- wrap some long lines

dougbu committed:
Move .0 package version workaround into Versions.props
- touch up SharedFramework.External.props

dougbu committed:
Enable Roslyn reference assemblies
- exclude ref/ assembly from packages other than targeting pack
- update Microsoft.AspNetCore.App.Ref.csproj
    - `%(IsReferenceAssembly)` and `%(ReferenceGrouping)` metadata no longer relevant
    - only ref/ assemblies are in `@(ReferencePathWithRefAssemblies)` item group

nits:
- remove now-unnecessary workaround
    - issues with TFM transition are behind us
- clean up Microsoft.AspNetCore.App.Runtime.csproj slightly
    - use `GeneratePathProperty="true"`
        - reorder item / property settings for meta-expansion
    - correct spelling errors and phrasing in comments

dougbu committed:
Update documentation to reflect recent changes
- remove CrossRepoBreakingChanges.md; was tied to old TeamCity infrastructure
  - also much less relevant given repo merges
- adjust details and examples in ReferenceResolution.md
  - reflect repo merges, Dependencies.props changes, and current Maestro++ channels
  - add a few more details e.g. specific files where Version.Details.xml versions are used

Tratcher commented:
@dougbu you are my new favorite person 😁. 

Can /docs/ReferenceAssemblies.md be removed?

Fixes https://github.com/dotnet/aspnetcore-internal/issues/2693

dougbu committed:
!fixup! Remove another irrelevant doc file

dougbu commented:
> Can /docs/ReferenceAssemblies.md be removed?

Yes 😺

wtgodbe commented on eng/SharedFramework.External.props:
```diff
@@ -7,6 +7,8 @@
 <Project>
 
   <!-- For the targeting pack, always use packages with PatchVersion=0 -->
+  <!-- !!! Need to get ref/ assemblies for these packages some other way in order to feed correct information into Roslyn. !!! -->
```
Why the exclamation points - was this a comment to yourself that you removed later?

dougbu replied:
Yes, I removed this later.

wtgodbe commented on Directory.Build.props:
```diff
@@ -130,7 +126,7 @@
     <TargetingPackInstallerBaseName>aspnetcore-targeting-pack</TargetingPackInstallerBaseName>
 
-    <!-- Produce targeting pack installers/packages once per major.minor. -->
+    <!-- Used to only produce targeting pack installers/packages once per major.minor. -->
```
nit - change to "This is used to..." to avoid devs accidentally interpreting this as "In the past, we only produced..."

wtgodbe commented on eng/SharedFramework.External.props:
```diff
@@ -9,7 +9,7 @@
   <!-- !!! Need to get ref/ assemblies for these packages some other way in order to feed correct information into Roslyn. !!! -->
   <!-- !!! For example, might update target paths to contain latest implementation assemblies after building /everything/ against the original package content. !!! -->
-  <PropertyGroup Condition="'$(IsServicingBuild)' == 'true' AND ('$(MSBuildProjectName)' == 'Microsoft.AspNetCore.App.Ref' OR '$(IsReferenceAssemblyProject)' == 'true')">
+  <PropertyGroup Condition=" '$(IsServicingBuild)' == 'true' AND '$(MSBuildProjectName)' == 'Microsoft.AspNetCore.App.Ref' ">
```
We can probably just move this `PropertyGroup` directly into the ref-pack's project, now that it's only used there

dougbu replied:
It's used everywhere now.

wtgodbe commented on eng/targets/ResolveReferences.targets:
```diff
@@ -77,7 +78,8 @@
     <_UnusedProjectReferenceProvider Include="@(ProjectReferenceProvider)" Exclude="@(Reference)" />
 
-    <_CompilationOnlyReference Condition="'$(TargetFramework)' == 'netstandard2.0'" Include="@(Reference->WithMetadataValue('NuGetPackageId','NETStandard.Library'))" />
+    <_CompilationOnlyReference Condition="'$(TargetFramework)' == 'netstandard2.0'"
```
Why move the conditions to a separate line for other items/properties, but keep the condition on the same line here?

dougbu replied:
Mostly depended on line length and where my cursor was when I hit **Enter** 😈 I don't particularly care whether the `Include` or the `Condition` appears first or second. But, if you do, lemme know.

wtgodbe commented on eng/targets/ResolveReferences.targets:
```diff
@@ -246,20 +238,20 @@
-    <Error Condition=" '$(IsAspNetCoreApp)' == 'true' AND '$(IsImplementationProject)' == 'true' AND '$(HasReferenceAssembly)' != 'true'" Text="All assemblies which have set IsAspNetCoreApp=true should produce a reference assembly." />
+    <Error
+        Condition=" '$(IsAspNetCoreApp)' == 'true' AND '$(IsImplementationProject)' == 'true' AND
+            !$(HasReferenceAssembly) AND '$(TargetFramework)' == '$(DefaultNetCoreTargetFramework)' "
```
How/where do we compute `HasReferenceAssembly` now?

wtgodbe commented on eng/Dependencies.props:
```diff
@@ -15,173 +15,193 @@ and are generated based on the last package release.
+
+    <!-- Get versions. -->
+    <LatestPackageReference Update="@(LatestPackageReference)">
+      <Version>$(%(VersionName)PackageVersion)</Version>
```
This scares me a _little_ bit - how strong is the guarantee that package version names will always follow this scheme? I believe darc/maestro require it (CC @mmitche), but it seems like we could easily break it for manual version updates like xunit.

dougbu replied:
If we break it, `@(Reference)` items fail loudly.

wtgodbe commented on eng/Dependencies.props:
```diff
@@ -15,173 +15,193 @@ and are generated based on the last package release.
+      <RTMVersion>$(%(VersionName)V0PackageVersion)</RTMVersion>
+    </LatestPackageReference>
+    <LatestPackageReference Update="@(LatestPackageReference->WithMetadataValue('Version', ''))">
+      <Version
```
Add a comment explaining why these 2 need to be done manually?

wtgodbe commented on eng/Versions.props:
```diff
@@ -154,6 +154,31 @@
+    !!! List below is likely incomplete. Should include all dotnet/runtime packages containing assemblies in
+    !!! Microsoft.AspNetCore.App.Ref or Microsoft.AspNetCore.App.Runtime.*.
+  -->
+  <PropertyGroup Condition=" '$(IsServicingBuild)' == 'true' ">
```
Why did you remove `AND '$(MSBuildProjectName)' == 'Microsoft.AspNetCore.App.Ref'` from the condition? We still want to reference the latest everywhere _except_ the ref pack

dougbu replied:
Because the condition was incorrect in 'master'. The 'release/3.1' condition is closer to what's needed: `$(MSBuildProjectName)' == 'Microsoft.AspNetCore.App.Ref' OR '$(IsReferenceAssemblyProject)' == 'true'`. But, that's incomplete because we split the ref/ and implementation versions in that branch.

I could conditionally compile implementation projects against the assemblies in these packages and test against the latest versions. But, I'm not sure that's worthwhile. Thoughts?

JunTaoLuo replied:
Why would this not be worthwhile?

dougbu replied:
It's worthwhile We discussed this offline and need to invoke the gods: @Pilchie and @ericstj 😺 

We had issues in the past because the updated packages from dotnet/corefx and dotnet/core-setup contained ref/ assemblies with newer assembly versions. This meant the ref/ assemblies built in dotnet/aspnetcore _required_ the newer ref/ assemblies when compiling and that wasn't our intent. @ericstj will newer ref/ assemblies change assembly versions in 5.0 servicing of dotnet/runtime❔

@Pilchie please confirm our requirements here are to
1. include the 5.0.0 ref/ assemblies from dotnet/runtime in our targeting packs
2. include any updated 5.0.x implementation assemblies from dotnet/runtime  in our runtime packs
3. compile against the 5.0.0 ref/ assemblies from dotnet/runtime
4. test against the 5.0.x implementation assemblies from dotnet/runtime

If that's correct, I can try referencing both packages, using `Compile` assets from the 5.0.0 packages and everything else from newer 5.0.x packages. But, I'm doubtful that will work correctly up the project dependency tree in this repo. I'm also not sure how to test this works correctly without dotnet/runtime and us doing a servicing exercise. Suggestions❔

Pilchie replied:
Those requirements seem correct from me.

wtgodbe replied:
It seems like it should be possible/supported (by roslyn) to always build against the oldest and run against the latest - maybe @agocke or someone else from Roslyn knows how we can accomplish that?

Pilchie replied:
That's not a question of the compiler, it's a question of the package contents, and msbuild targets that locate things.  The compiler just gets a path on disk to the thing to reference.

ericstj replied:
> @ericstj will newer ref/ assemblies change assembly versions in 5.0 servicing of dotnet/runtime

Ref assemblies that are part of the ref-pack will never change in servicing.  Ref assemblies in the transport package we provide for extensions should also not change.

Individual packages will change as needed so that they can have unique assembly versions.

dougbu replied:
🆗 @ericstj that means I need to hack things to get our projects compiling against 5.0.0 versions of the assemblies and running against 5.0.x for this specific set of packages. I haven't tried my "reference both versions" idea yet and still have no idea how to test it out without a dotnet/runtime / dotnet/aspnetcore servicing exercise.

ericstj replied:
Wasn’t this what was already happening in 3.x?  I thought that’s exactly what we made work for the ref pack servicing.  Let me know if there is something different we should be doing to make things easier.

dougbu replied:
Yes, we're dealing with similar incoming packages in 3.x but aren't using Roslyn to generate the ref/ assemblies. That means it's _much_ easier in 3.x to compile our ref/ assemblies against the 3.1.0 ref/ assemblies from outside aspnetcore. We only need to restrict ourselves to the 3.1.0 versions of these packages when building ref/ projects and Microsoft.AspNetCore.App.Ref. Further, we were already in 3.x servicing when we found the problem which made testing simpler.

With this change, we're building the ref/ and implementation projects with the same `@(Reference)` items and no longer have separate ref/ projects. That makes testing against the 5.0.x packages more complicated and gnarly to test.

dougbu replied:
Oops, "ref/ and implementation **assemblies**"

wtgodbe commented on Directory.Build.targets:
```diff
@@ -107,8 +107,13 @@
     <IsProjectReferenceProvider Condition=" '$(IsProjectReferenceProvider)' == '' AND '$(IsImplementationProject)' == 'true' AND '$(PackAsTool)' != 'true' ">true</IsProjectReferenceProvider>
 
     <HasReferenceAssembly
```
Ah, here it is

wtgodbe commented on src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj:
```diff
@@ -122,7 +125,10 @@ This package is an internal implementation of the .NET Core SDK and is not meant
     </Reference>
 
-    <Reference Include="microsoft.netcore.app.runtime.$(RuntimeIdentifier)" ExcludeAssets="All" PrivateAssets="All" />
+    <Reference Include="microsoft.netcore.app.runtime.$(RuntimeIdentifier)"
```
nit - make capitalization consistent with other references to `m.nc.app`

wtgodbe commented on docs/ReferenceResolution.md:
```diff
@@ -65,27 +68,28 @@ Steps for adding a new package dependency to an existing project. Let's say I'm
-        dotnet/CoreFx dependency version should be determined based on the build that produced the chosen
-        Microsoft.NETCore.App. That is, the dotnet/CoreFx dependency and Microsoft.NETCore.App should be
-        coherent.
+        dotnet/runtime dependency version for System.Banana should be determined based on the dotnet/aspnetcore build
```
aspnetcore -> aspnetcore-tooling

wtgodbe replied:
Actually, we don't depend on tooling anymore - I think we can get rid of all CPDs in this repo now that the only repo below us is dotnet/runtime (@mmitche please confirm)

dougbu replied:
No; runtime bits flow into aspnetcore and from there into aspnetcore-tooling. This part of the document describes the CPD setup needed to keep that coherent.

JunTaoLuo replied:
Yes language here is correct.

wtgodbe commented on docs/ReferenceResolution.md:
```diff
@@ -65,27 +68,28 @@ Steps for adding a new package dependency to an existing project. Let's say I'm
-        Microsoft.NETCore.App. That is, the dotnet/CoreFx dependency and Microsoft.NETCore.App should be
-        coherent.
+        dotnet/runtime dependency version for System.Banana should be determined based on the dotnet/aspnetcore build
+        that produced the chosen Microsoft.CodeAnalysis.Razor. That is, the dotnet/runtime and dotnet/aspnetcore
```
Same here

dougbu replied:
As above

wtgodbe commented:
I must have missed it - where/how do we actually hook in to the roslyn `ref` project generation?

dougbu commented:
> where/how do we actually hook in to the roslyn `ref` project generation?

It's a single property setting [here](https://github.com/dotnet/aspnetcore/pull/23403/files#diff-56f8482aeb786dd105fbf93dedf2615fR112).

JunTaoLuo commented on eng/targets/ResolveReferences.targets:
```diff
@@ -246,20 +238,20 @@
+        Condition=" '$(IsAspNetCoreApp)' == 'true' AND '$(IsImplementationProject)' == 'true' AND
+            !$(HasReferenceAssembly) AND '$(TargetFramework)' == '$(DefaultNetCoreTargetFramework)' "
+        Text="All assemblies which have set IsAspNetCoreApp=true should produce a reference assembly for default TFM." />
+    <Warning Condition=" '$(IsAspNetCoreApp)' == 'true' AND '$(IsImplementationProject)' != 'true' "
```
Why isn't this an error. Are there actually cases of this being valid? We could add an option to opt out of the error but by default I feel this is wrong enough to be an error.

dougbu replied:
This and the task below were `Warning`s when I added them because I was worried about the Blazor / Mono ref/ assembly project that's since been removed. Will make them `Error`s.

JunTaoLuo commented on eng/targets/ResolveReferences.targets:
```diff
@@ -246,20 +238,20 @@
+        Text="All assemblies which have set IsAspNetCoreApp=true should produce a reference assembly for default TFM." />
+    <Warning Condition=" '$(IsAspNetCoreApp)' == 'true' AND '$(IsImplementationProject)' != 'true' "
+        Text="Only implementation projects should set IsAspNetCoreApp=true." />
+    <Warning Condition=" '$(IsAspNetCoreApp)' != 'true' AND $(HasReferenceAssembly) "
```
Ditto

wtgodbe reviewed (changes requested):
Everything looks good except for the question about building against old refs & running against new impl's - will approve once that's resolved

dougbu committed:
!fixup! Address PR review suggestions
- convert a couple of warnings to errors
- use consistent casing for Microsoft.NETCore.App.Runtime.* packages
- reduce `%(LatestPackageReference.Version)` metadata special cases
- add and improve comments e.g.
    - improve comments about `$(*V0PackageVersion)` properties
    - improve placement of comments about item removal in ResolveReferences.targets
    - confirmed `$(*V0PackageVersion)` property list is complete

nits:
- fix solution example in ReferenceResolution.md
- remove item group definition for `@(LatestPackageReference)`
- remove `%(LatestPackageReference.VersionName)` metadata after use; large item group
    - similarly, remove `%(LatestPackageReference.RTMVersion)` when not needed; just complicates `Condition`s

When I squash, I must remember this fixes
- #14801
- dotnet/aspnetcore-internal#2693

dougbu committed:
Actually use `%(LatestPackageReference.RTMVersion)` metadata
- gather RTM package references in a new project
    - a (very) separate project to work around package conflict resolution
    - empty `Test` target works around Arcade's testing approach
- new target in ResolveReferences.targets updates relevant assembly paths to use the RTM packages
    - done as soon as possible after `ResolvePackageAssets` determines the paths
    - done for all compilation inputs, not just ref/ assemblies

dougbu commented on docs/ReferenceResolution.md:
```diff
@@ -29,30 +31,31 @@ The requirements that led to this system are:
 1. Create the .csproj
 2. Run `eng/scripts/GenerateProjectList.ps1`
-3. Add it to Extensions.sln
+3. Add it to Mvc.sln (for example)
```
Note to self: Make this AspNetCore.sln

dougbu commented on eng/Versions.props:
```diff
@@ -157,6 +155,35 @@
+    <SystemWindowsExtensionsV0PackageVersion>$(SystemWindowsExtensionsPackageVersion.Split('.')[0]).$(SystemWindowsExtensionsPackageVersion.Split('.')[1]).0</SystemWindowsExtensionsV0PackageVersion>
+  </PropertyGroup>
+
+  <!-- Fake a servicing build by using the previous version of a couple of the above packages for ref/ assemblies. -->
+  <PropertyGroup Label="Test only; remove before merging !!!">
+    <SystemIOPipelinesV0PackageVersion>5.0.0-preview.8.20354.5</SystemIOPipelinesV0PackageVersion>
+    <SystemSecurityCryptographyCngV0PackageVersion>5.0.0-preview.8.20354.5</SystemSecurityCryptographyCngV0PackageVersion>
+  </PropertyGroup>
```
/FYI this block is about my earlier comment:
> still have no idea how to test it out without a dotnet/runtime / dotnet/aspnetcore servicing exercise.

Idea is to pretend these two packages previous versions are the RTM version and we're at RTM+1. If this works, I'll remove the "test only" commit and let everyone who wants to review know.

dougbu commented on eng/Versions.props:
```diff
@@ -189,7 +216,7 @@
     <MicrosoftCodeAnalysisCSharpPackageVersion>3.7.0-4.20351.7</MicrosoftCodeAnalysisCSharpPackageVersion>
     <MicrosoftCodeAnalysisCSharpWorkspacesPackageVersion>3.7.0-4.20351.7</MicrosoftCodeAnalysisCSharpWorkspacesPackageVersion>
-    <MicrosoftCodeAnalysisFxCopAnalyzersVersion>3.0.0</MicrosoftCodeAnalysisFxCopAnalyzersVersion>
+    <MicrosoftCodeAnalysisFxCopAnalyzersPackageVersion>3.0.0</MicrosoftCodeAnalysisFxCopAnalyzersPackageVersion>
```
Change here and at line 255 is about property name consistency. If these properties were automatic, eng/scripts/CodeCheck.ps1 would currently fail; it requires `$(*PackageVersion)` properties for everything in eng/Version.Details.xml. It appears we carried that consistency over to the manual version properties.

If someone (@Pilchie this was your addition) wants to remove `Package` from the names, we can certainly do that -- separately. File an issue and put it on in DoI or Build project. Who knows (@mmitche❔) such a change might even reduce the size of our binlogs or speed builds up a bit 😺

ericstj commented:
> increases the targeting pack size slightly

One idea we had in dotnet/runtime was to use the ILLinker to trim the resulting reference assemblies that are shipped in the ref pack.  We already do this for our runtime assemblies, doing so for ref might be a good use of linker.  The settings we use with it to root public API and trim unreachable code are here: https://github.com/dotnet/runtime/blob/69b0d160953f5c920e52f021366743623693c158/eng/illink.targets#L192

dougbu commented:
> One idea we had in dotnet/runtime was to use the ILLinker to trim the resulting reference assemblies that are shipped in the ref pack.

That sounds like fun 😺 but I'm not that sure the small size bump merits the extra work. My latest build produced a 2.02 MB Microsoft.AspNetCore.App.Ref.5.0.0-ci.nupkg while the one from the most recent rolling build of master was 1.74 MB. Thoughts @Pilchie❔

Pilchie commented:
I suspect it might be a fair bit of work, given that our reference assemblies need to have some internals so that test assemblies can be built that have `InternalsVisibleTo`.  It seems like a significant thing to manage for a negligible benefit.

dougbu commented on eng/targets/ResolveReferences.targets:
```diff
@@ -192,24 +190,52 @@
-    <Error Condition="'$(TargetFrameworkIdentifier)' != '.NETFramework' AND '%(Reference.Identity)' != '' AND ! Exists('%(Reference.Identity)') AND '$(DisablePackageReferenceRestrictions)' != 'true'"
-           Code="MSB3245"
-           Text="Could not resolve this reference. Could not locate the package or project for &quot;%(Reference.Identity)&quot;. Did you update baselines and dependencies lists? See docs/ReferenceResolution.md for more details." />
+    !!! Note lib/ assemblies chosen for (say) .netstandard2.0 compilation are also be updated. Thoughts? !!!
```
I believe this is the **Right Thing:tm:** to do but would appreciate confirmation from @Pilchie and @ericstj.

dougbu replied:
Confirmed offline w/ @Pilchie. We're good to go here.

I'll remove the `!!!` and change the comment to mention the main scenario that would break this target: Layout changes to one of the dozen dotnet/runtime packages could mean we pick a compatible assembly from the new package and the path doesn't exist in the old one. Fortunately, this will break the build with complaints about the invalid strong name. I guess the reverse is also true but can't imagine us _removing_ a TFM from a package during servicing.

dougbu reviewed (commented):
🆗 I declare this PR working.

The https://dev.azure.com/dnceng/public/_build/results?buildId=731663 build might still time out (unrelated to my changes) but otherwise did all the **Right Thing:tm:s** Have confirmed the downlevel assemblies landed in Microsoft.AspNetCore.App.Ref.5.0.0-ci.nupkg and shipped assemblies all built against what they should have e.g. Microsoft.AspNetCore.Http.Features.dll references the expected (older) ref/ assembly for System.IO.Pipelines.
``` text
System.Collections, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
System.ComponentModel, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
System.IO.Pipelines, Version=5.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51
System.Linq, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
System.Net.Primitives, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
```

I'll clean up the commits and remove the "test only" parts in preparation for a final review. In the meantime, I would appreciate comments on below `!!!` bit since I'd rather not check that question in 😈

dougbu commented:
@Pilchie 
> I suspect it might be a fair bit of work

Maybe not quite as much as you suspect. My objection was based on the size of the target @ericstj pointed us to and the potential build perf degradation due to a lot of `<Exec>` invocations. If we did anything, it would be in the Microsoft.AspNetCore.App.Ref project and should not affect anything else building in this repo.

ericstj commented:
Fair enough.  FWIW we saw a lot larger size differences when we were looking at roslyn's references.  For example, I spot checked System.Text.Json right now and our ref is 25KB vs Roslyn's which is 75KB.  The nice thing about the linker is that it can solve the internals problem for you.  IIRC we were trying idefs and this was hard because the compiler needs legit syntax, whereas with the linker you can root just the internals you need and it will walk the type graph.

jaredpar commented:
>  FWIW we saw a lot larger size differences when we were looking at roslyn's references. For example, I spot checked System.Text.Json right now and our ref is 25KB vs Roslyn's which is 75KB

This is to some degree expected. The reference assemblies generated by the runtime team are omitting members. This can be done in runtime because they have a good grasp of their specific domain and can safely omit some types / members. Though this has bit them before and ended up causing real world customer issues (the stripping of `struct` fields that were `private` as an example). 

The general compiler feature though must produce reference assemblies that are correct in any customer domain. That means we'll generally include more data because it's a superset of what runtime would produce.

dougbu commented:
:up::date: to consolidate changes since @wtgodbe and @JunTaoLuo reviewed (equivalent of https://github.com/dotnet/aspnetcore/pull/23403/commits/c65704f8782fe331080a17b3f0ea28c5072ae6a8) into two commits, remove the test-only bits, and rebase on preview8 branch. Please review or re-review in case there's anything I missed.

@Pilchie we didn't chat about this but I chose to label this a tell-mode change. LMK if that's not right.

dougbu commented:
The shift-click menu doesn't always work. The URI for the last 9 commits (all but removing the ref/ projects) is https://github.com/dotnet/aspnetcore/pull/23403/files/32dd707601f5b832060e8bbe49be5197c844545f..7ebf486c9a29c462ad8956d117b224c231111d53 at the moment.

wtgodbe commented on eng/RTMVersions/RTMVersions.csproj:
```diff
@@ -0,0 +1,28 @@
+    in all servicing builds. Cannot reference two versions of a package, mandating this separate package.
+  -->
+  <PropertyGroup>
+    <TargetFramework>net5.0</TargetFramework>
```
Can't be parameterized?

dougbu replied:
Not unless I remove the `Directory.Build.*` files. I'll test out whether that causes any problems after getting this in and doing the merge to 'master'. It's no worse at the moment than the hard-coded TFMs in (say) our project templates. But, I'll clean it up next if possible.

Pilchie commented:
I think tell mode is fine - I'll explicitly mention it in tactics, since it *is* a pretty significant change.

Pilchie commented:
Yipee!!
